numpy/core/tests/test_numeric.py

```diff
@@ -2877,6 +2877,16 @@ def test_zero_size(self):
         with pytest.raises(ValueError):
             np.correlate(np.ones(1000), np.array([]), mode='full')
 
+    def test_mode(self):
+        d = np.ones(100)
+        k = np.ones(3)
+        default_mode = np.correlate(d, k, mode='valid')
+        with assert_warns(DeprecationWarning):
+            valid_mode = np.correlate(d, k, mode='v')
+        assert_array_equal(valid_mode, default_mode)
+        # integer mode
+        assert_array_equal(np.correlate(d, k, mode=0), valid_mode)
```

Comment:
Can you add a test of an invalid integer too?

Also, it may make sense to add a test here too:

https://github.com/numpy/numpy/blob/aeb23741b1186e6ae525c4dd7f46ce9e1378e670/numpy/core/src/multiarray/_multiarray_tests.c.src#L2088-L2101

https://github.com/numpy/numpy/blob/aeb23741b1186e6ae525c4dd7f46ce9e1378e670/numpy/core/tests/test_conversion_utils.py#L163-L174

Reply:
I created the tests, but I can't seem to find a way to make `PyArray_CorrelatemodeConverter` function visible like `PyArray_ClipmodeConverter` in this file:  https://github.com/numpy/numpy/blob/aeb23741b1186e6ae525c4dd7f46ce9e1378e670/numpy/core/src/multiarray/_multiarray_tests.c.src#L2088-L2101 

If I directly `#include "conversion_utils.h"` it throws a bunch of errors while building, which is understandable.<br>I think  `PyArray_ClipmodeConverter` is visible to this file as it is a global C API.
Or is there something I'm missing?

Reply:
> I think `PyArray_ClipmodeConverter` is visible to this file as it is a global C API.

Hmm, that's possible.

> If I directly #include "conversion_utils.h" it throws a bunch of errors while building, which is understandable.

Can you give an example of those errors?

Reply:
Here's the full traceback, it seems like a compiler issue?
```
In file included from numpy/core/include/numpy/ndarrayobject.h:21:0,
                 from numpy/core/include/numpy/arrayobject.h:4,
                 from numpy/core/src/multiarray/_multiarray_tests.c.src:5:
numpy/core/include/numpy/__multiarray_api.h:1090:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, PyArray_Dims *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:7:1: note: in expansion of macro ‘PyArray_IntpConverter’
 PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq);
 ^~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1091:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[176])
          ^
numpy/core/src/multiarray/conversion_utils.h:7:1: note: in expansion of macro ‘PyArray_IntpConverter’
 PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq);
 ^~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1093:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, PyArray_Chunk *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:13:1: note: in expansion of macro ‘PyArray_BufferConverter’
 PyArray_BufferConverter(PyObject *obj, PyArray_Chunk *buf);
 ^~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1094:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[177])
          ^
numpy/core/src/multiarray/conversion_utils.h:13:1: note: in expansion of macro ‘PyArray_BufferConverter’
 PyArray_BufferConverter(PyObject *obj, PyArray_Chunk *buf);
 ^~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1099:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, npy_bool *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:16:1: note: in expansion of macro ‘PyArray_BoolConverter’
 PyArray_BoolConverter(PyObject *object, npy_bool *val);
 ^~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1100:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[179])
          ^
numpy/core/src/multiarray/conversion_utils.h:16:1: note: in expansion of macro ‘PyArray_BoolConverter’
 PyArray_BoolConverter(PyObject *object, npy_bool *val);
 ^~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1102:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, char *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:19:1: note: in expansion of macro ‘PyArray_ByteorderConverter’
 PyArray_ByteorderConverter(PyObject *obj, char *endian);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1103:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[180])
          ^
numpy/core/src/multiarray/conversion_utils.h:19:1: note: in expansion of macro ‘PyArray_ByteorderConverter’
 PyArray_ByteorderConverter(PyObject *obj, char *endian);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1126:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, NPY_SORTKIND *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:22:1: note: in expansion of macro ‘PyArray_SortkindConverter’
 PyArray_SortkindConverter(PyObject *obj, NPY_SORTKIND *sortkind);
 ^~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1127:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[188])
          ^
numpy/core/src/multiarray/conversion_utils.h:22:1: note: in expansion of macro ‘PyArray_SortkindConverter’
 PyArray_SortkindConverter(PyObject *obj, NPY_SORTKIND *sortkind);
 ^~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1177:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, void *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:25:1: note: in expansion of macro ‘PyArray_SearchsideConverter’
 PyArray_SearchsideConverter(PyObject *obj, void *addr);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1178:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[205])
          ^
numpy/core/src/multiarray/conversion_utils.h:25:1: note: in expansion of macro ‘PyArray_SearchsideConverter’
 PyArray_SearchsideConverter(PyObject *obj, void *addr);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:862:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:28:1: note: in expansion of macro ‘PyArray_PyIntAsInt’
 PyArray_PyIntAsInt(PyObject *o);
 ^~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:863:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[100])
          ^
numpy/core/src/multiarray/conversion_utils.h:28:1: note: in expansion of macro ‘PyArray_PyIntAsInt’
 PyArray_PyIntAsInt(PyObject *o);
 ^~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:865:22: error: expected declaration specifiers or ‘...’ before ‘*’ token
         (*(npy_intp (*)(PyObject *)) \
                      ^
numpy/core/src/multiarray/conversion_utils.h:31:1: note: in expansion of macro ‘PyArray_PyIntAsIntp’
 PyArray_PyIntAsIntp(PyObject *o);
 ^~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:865:24: error: expected ‘)’ before ‘(’ token
         (*(npy_intp (*)(PyObject *)) \
                        ^
numpy/core/src/multiarray/conversion_utils.h:31:1: note: in expansion of macro ‘PyArray_PyIntAsIntp’
 PyArray_PyIntAsIntp(PyObject *o);
 ^~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:866:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[101])
          ^
numpy/core/src/multiarray/conversion_utils.h:31:1: note: in expansion of macro ‘PyArray_PyIntAsIntp’
 PyArray_PyIntAsIntp(PyObject *o);
 ^~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1063:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, npy_intp *, int)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:37:1: note: in expansion of macro ‘PyArray_IntpFromSequence’
 PyArray_IntpFromSequence(PyObject *seq, npy_intp *vals, int maxvals);
 ^~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1064:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[167])
          ^
numpy/core/src/multiarray/conversion_utils.h:37:1: note: in expansion of macro ‘PyArray_IntpFromSequence’
 PyArray_IntpFromSequence(PyObject *seq, npy_intp *vals, int maxvals);
 ^~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1081:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(int, int)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:40:1: note: in expansion of macro ‘PyArray_TypestrConvert’
 PyArray_TypestrConvert(int itemsize, int gentype);
 ^~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1082:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[173])
          ^
numpy/core/src/multiarray/conversion_utils.h:40:1: note: in expansion of macro ‘PyArray_TypestrConvert’
 PyArray_TypestrConvert(int itemsize, int gentype);
 ^~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1150:21: error: expected ‘)’ before ‘*’ token
         (*(PyObject * (*)(int, npy_intp const *)) \
                     ^
numpy/core/src/multiarray/conversion_utils.h:43:1: note: in expansion of macro ‘PyArray_IntTupleFromIntp’
 PyArray_IntTupleFromIntp(int len, npy_intp const *vals);
 ^~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1151:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[196])
          ^
numpy/core/src/multiarray/conversion_utils.h:43:1: note: in expansion of macro ‘PyArray_IntTupleFromIntp’
 PyArray_IntTupleFromIntp(int len, npy_intp const *vals);
 ^~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1444:12: error: expected identifier or ‘(’ before ‘int’
         (*(int (*)(PyObject *, NPY_SELECTKIND *)) \
            ^
numpy/core/src/multiarray/conversion_utils.h:49:1: note: in expansion of macro ‘PyArray_SelectkindConverter’
 PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
numpy/core/include/numpy/__multiarray_api.h:1445:10: error: expected ‘)’ before ‘PyArray_API’
          PyArray_API[298])
          ^
numpy/core/src/multiarray/conversion_utils.h:49:1: note: in expansion of macro ‘PyArray_SelectkindConverter’
 PyArray_SelectkindConverter(PyObject *obj, NPY_SELECTKIND *selectkind);
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~

```

Reply:
Yeah, looks like your diagnosis is correct; `multiarray_tests` requires things to be in the public API. I think it's ok to leave things as is for now, thanks for investigating